File: hawk/src/test/java/com/orhanobut/hawk/SharedPreferencesStorageTest.java
package com.orhanobut.hawk;

import android.content.SharedPreferences;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import static com.google.common.truth.Truth.assertThat;
import static junit.framework.Assert.fail;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
@Config(manifest = Config.NONE)
public class SharedPreferencesStorageTest {

  SharedPreferencesStorage storage;
  SharedPreferences.Editor editor;
  SharedPreferences preferences;

  @Before public void setup() {
    editor = spy(new SharedPreferences.Editor() {
      @Override public SharedPreferences.Editor putString(String key, String value) {
        return this;
      }

      @Override public SharedPreferences.Editor putStringSet(String key, Set<String> values) {
        return this;
      }

      @Override public SharedPreferences.Editor putInt(String key, int value) {
        return this;
      }

      @Override public SharedPreferences.Editor putLong(String key, long value) {
        return this;
      }

      @Override public SharedPreferences.Editor putFloat(String key, float value) {
        return this;
      }

      @Override public SharedPreferences.Editor putBoolean(String key, boolean value) {
        return this;
      }

      @Override public SharedPreferences.Editor remove(String key) {
        return this;
      }

      @Override public SharedPreferences.Editor clear() {
        return this;
      }

      @Override public boolean commit() {
        return true;
      }

      @Override public void apply() {
      }
    });
    preferences = spy(new SharedPreferences() {
      @Override public Map<String, ?> getAll() {
        Map<String, Object> map = new HashMap<>();
        map.put("key", "value");
        return map;
      }

      @Override public String getString(String key, String defValue) {
        return null;
      }

      @Override public Set<String> getStringSet(String key, Set<String> defValues) {
        return null;
      }

      @Override public int getInt(String key, int defValue) {
        return 0;
      }

      @Override public long getLong(String key, long defValue) {
        return 0;
      }

      @Override public float getFloat(String key, float defValue) {
        return 0;
      }

      @Override public boolean getBoolean(String key, boolean defValue) {
        return false;
      }

      @Override public boolean contains(String key) {
        return false;
      }

      @Override public Editor edit() {
        return editor;
      }

      @Override public void registerOnSharedPreferenceChangeListener(OnSharedPreferenceChangeListener listener) {

      }

      @Override public void unregisterOnSharedPreferenceChangeListener(OnSharedPreferenceChangeListener listener) {

      }
    });
    storage = new SharedPreferencesStorage(preferences);
  }

  @Test public void putObject() throws Exception {
    storage.put("key", "value");

    verify(preferences).edit();
    verify(editor).putString("key", "value");
    verify(editor).commit();
  }

  @Test public void throwExceptionOnNullKeysPassedOnPut() {
    try {
      storage.put(null, "value");
      fail("key should not be null");
    } catch (Exception e) {
      assertThat(e).hasMessage("key should not be null");
    }
  }

  @Test public void get() throws Exception {
    storage.get("key");

    verify(preferences).getString("key", null);
  }

  @Test public void remove() throws Exception {
    storage.delete("key");

    verify(preferences).edit();
    verify(editor).remove("key");
    verify(editor).commit();
  }

  @Test public void contains() throws Exception {
    storage.contains("key");

    verify(preferences).contains("key");
  }

  @Test public void clear() throws Exception {
    storage.clear();

    verify(preferences).edit();
    verify(editor).clear();
    verify(editor).commit();
  }

  @Test public void count() throws Exception {
    long count = storage.count();

    verify(preferences).getAll();
    assertThat(count).isEqualTo(1);
  }
}

File: hawk/src/test/java/com/orhanobut/hawk/SharedPreferencesStorageTest.kt
package com.orhanobut.hawk

import android.content.SharedPreferences

import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

import java.util.HashMap

import com.google.common.truth.Truth.assertThat
import junit.framework.Assert.fail
import org.mockito.Mockito.spy
import org.mockito.Mockito.verify

@RunWith(RobolectricTestRunner::class)
class SharedPreferencesStorageTest {

  private lateinit var storage: SharedPreferencesStorage
  private lateinit var editor: SharedPreferences.Editor
  private lateinit var preferences: SharedPreferences

  @Before fun setup() {
    editor = spy<SharedPreferences.Editor>(object : SharedPreferences.Editor {
      override fun putString(key: String, value: String?): SharedPreferences.Editor {
        return this
      }

      override fun putStringSet(key: String, values: Set<String>?): SharedPreferences.Editor {
        return this
      }

      override fun putInt(key: String, value: Int): SharedPreferences.Editor {
        return this
      }

      override fun putLong(key: String, value: Long): SharedPreferences.Editor {
        return this
      }

      override fun putFloat(key: String, value: Float): SharedPreferences.Editor {
        return this
      }

      override fun putBoolean(key: String, value: Boolean): SharedPreferences.Editor {
        return this
      }

      override fun remove(key: String): SharedPreferences.Editor {
        return this
      }

      override fun clear(): SharedPreferences.Editor {
        return this
      }

      override fun commit(): Boolean {
        return true
      }

      override fun apply() {}
    })
    preferences = spy<SharedPreferences>(object : SharedPreferences {
      override fun getAll(): Map<String, *> {
        val map = HashMap<String, Any>()
        map["key"] = "value"
        return map
      }

      override fun getString(key: String, defValue: String?): String? {
        return null
      }

      override fun getStringSet(key: String, defValues: Set<String>?): Set<String>? {
        return null
      }

      override fun getInt(key: String, defValue: Int): Int {
        return 0
      }

      override fun getLong(key: String, defValue: Long): Long {
        return 0
      }

      override fun getFloat(key: String, defValue: Float): Float {
        return 0f
      }

      override fun getBoolean(key: String, defValue: Boolean): Boolean {
        return false
      }

      override fun contains(key: String): Boolean {
        return false
      }

      override fun edit(): SharedPreferences.Editor? {
        return editor
      }

      override fun registerOnSharedPreferenceChangeListener(listener: SharedPreferences.OnSharedPreferenceChangeListener) {

      }

      override fun unregisterOnSharedPreferenceChangeListener(listener: SharedPreferences.OnSharedPreferenceChangeListener) {

      }
    })
    storage = SharedPreferencesStorage(preferences)
  }

  @Test fun putObject() {
    storage.put("key", "value")

    verify(preferences).edit()
    verify(editor).putString("key", "value")
    verify(editor).commit()
  }

  @Test fun throwExceptionOnNullKeysPassedOnPut() {
    try {
      storage.put(null, "value")
      fail("key should not be null")
    } catch (e: Exception) {
      assertThat(e).hasMessage("key should not be null")
    }

  }

  @Test fun get() {
    storage.get<Any>("key")

    verify(preferences).getString("key", null)
  }

  @Test fun remove() {
    storage.delete("key")

    verify(preferences).edit()
    verify(editor).remove("key")
    verify(editor).commit()
  }

  @Test fun contains() {
    storage.contains("key")

    verify(preferences).contains("key")
  }

  @Test fun deleteAll() {
    storage.deleteAll()

    verify(preferences).edit()
    verify(editor).clear()
    verify(editor).commit()
  }

  @Test fun count() {
    val count = storage.count()

    verify(preferences).all
    assertThat(count).isEqualTo(1)
  }
}